File: src/javacodes/Array/NextPermutation.java
package javacodes.Array;

public class NextPermutation {

    static void swap(int[] nums, int i, int j) {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }

    static void nextPerm(int[] nums) {
        int n = nums.length;

        int mIndex = -1;
        for(int i = n - 2; i >= 0; i--) {
            if(nums[i] < nums[i+1]) {
                mIndex = i;
                int sSm = n - 1;
                while(nums[mIndex] >= nums[sSm]) {
                    sSm -= 1;
                }
                swap(nums, mIndex, sSm);
                break;
            }
        }

        // Reverse
        for (int i = mIndex + 1; i < (n + mIndex + 1) / 2; i++) {
            swap(nums, i, n + mIndex - i);
        }
    }

    public static void main(String[] args) {
        int[] arr = {1, 3, 5, 4, 2};
        nextPerm(arr);

        for(int a: arr) {
            System.out.print(a + " ");
        }
    }
}

File: src/pythoncodes/arrays/next_permutation.py
def swap(s, i, j):
    temp = s[i]
    s[i] = s[j]
    s[j] = temp


def next_permutation(nums):
    n = len(nums)
    mIndex = -1
    for i in range(n - 2, -1, -1):
        if nums[i] < nums[i + 1]:
            mIndex = i
            sm = n - 1
            while nums[mIndex] >= nums[sm]:
                sm -= 1
            swap(nums, mIndex, sm)
            print(nums)
            break

    for i in range(mIndex + 1, (n + mIndex + 1) // 2):
        swap(nums, i, n + mIndex - i)


if __name__ == '__main__':
    arr = [1, 3, 2, 4, 3]
    next_permutation(arr)

    print(arr)
